src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -716,19 +719,19 @@ private void CompletePipe()
                 // Return all segments
                 // if _readHead is null we need to try return _commitHead
                 // because there might be a block allocated for writing
-                BufferSegment segment = _readHead ?? _readTail;
-                while (segment != null)
-                {
-                    BufferSegment returnSegment = segment;
-                    segment = segment.NextSegment;
-
-                    returnSegment.ResetMemory();
-                }
+                current = _readHead ?? _readTail;
 
                 _writingHead = null;
                 _readHead = null;
                 _readTail = null;
             }
+
+            while (current != null)
+            {
+                BufferSegment next = current.NextSegment;
+                current.ResetMemory();
```

Comment:
I wonder if we should return segments here instead of letting them get GCed

Reply:
No point pooling as they won't get used as the Pipe is completed? (as its a per pipe pool currently)

Reply:
But pipe can be reset and I expect that's what we do for keep-alive requests.

Reply:
Ah! Ok :)

Reply:
Fixed